client/js/serverstate.js

```diff
@@ -430,7 +430,8 @@ function receiveStateFromServer(args) {
 
   mouseTarget = null;
   deltaID = args._meta.deltaID;
-  for(const widget of widgetFilter(w=>w.get('parent')===null))
+  const topSurface = $('#topSurface');
+  for(const widget of widgetFilter(w=>w.domElement.parentElement === topSurface))
```

Comment:
Each one of those removes its children recursively. IIRC, the leaves of the tree have to be removed first to avoid issues with already removed parents.

Reply:
Sorry, I wasn't clear. How do you know that the widgets removed are on the top surface because their original parent does not exist? That is, why doesn't this just clear the table?

Reply:
I'm still not quite following. This _is_ clearing the table because the server sent a complete new state.